### PR TITLE
core: ensure txindex will be triggered at least once

### PR DESCRIPTION
**Pull request by jsvisa**
Currently, we trigger the logic to (un)index transactions when the node receives a new block. However, in some cases the node may not receive new blocks (eg, when the Geth node is configured with `--nodiscovery`, or when it acts as an RPC node for historical only data). 
In these situations, the Geth node user may not have previously configured `txlookuplimit` (i.e. the default of around one year), but later he realizes they need to index all historical blocks. However, adding `txlookuplimit=0` and restarting geth has no effect.

So in this case, I think if we start a timer, there would at least be one opportunity to run the indexing process.

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -2394,9 +2394,25 @@ func (bc *BlockChain) maintainTxIndex() {
 	}
 	defer sub.Unsubscribe()
 
+	// If geth is running in read-only mode(not receiving new blocks),
```
Comment (rjl493456442) [inline core/blockchain.go:2397]: Instead of introducing a timer, we can always spin up a indexBlocks action before loop, as the initialization step.

In the scenarios you mentioned, Geth should index/uninindex the unfinished chain segments and then do nothing(because chain has no progress)

Comment (jsvisa) [reply core/blockchain.go:2397]: thanks for the advice.

Comment (jsvisa) [reply core/blockchain.go]: @rjl493456442 please take another look, thanks

File: core/blockchain.go [hunk 2]
```diff
@@ -2383,6 +2385,13 @@ func (bc *BlockChain) indexBlocks(tail *uint64, head uint64, done chan struct{})
 func (bc *BlockChain) maintainTxIndex() {
 	defer bc.wg.Done()
 
+	// Geth should always index/uninindex the unfinished chain segments,
+	// in case the chain has no progress for a long time.
+	head := rawdb.ReadHeadBlock(bc.db)
+	if head != nil {
+		bc.indexBlocks(rawdb.ReadTxIndexTail(bc.db), head.NumberU64(), nil)
+	}
+
```
Comment (rjl493456442) [inline core/blockchain.go:2394]: Perhaps this version is cleaner?

```diff
diff --git a/core/blockchain.go b/core/blockchain.go
index 3952c31b68..807da7691f 100644
--- a/core/blockchain.go
+++ b/core/blockchain.go
@@ -2394,6 +2394,14 @@ func (bc *BlockChain) maintainTxIndex() {
        }
        defer sub.Unsubscribe()
 
+       // Launch the initial processing if chain is not empty. This step is
+       // useful in the scenarios that chain has no progress, thus the indexer
+       // won't be triggered at all.
+       head := rawdb.ReadHeadBlock(bc.db)
+       if head != nil {
+               done = make(chan struct{})
+               bc.indexBlocks(rawdb.ReadTxIndexTail(bc.db), head.NumberU64(), done)
+       }
        for {
                select {
                case head := <-headCh:
diff --git a/tests/testdata b/tests/testdata
index ee3fa4c86d..0902578614 160000
--- a/tests/testdata
+++ b/tests/testdata
```

Comment (jsvisa) [reply core/blockchain.go]: thanks and lgtm

**Comment by jsvisa** [comment]
@rjl493456442 please take another look, thanks

File: core/blockchain.go [hunk 3]
```diff
@@ -2394,6 +2394,16 @@ func (bc *BlockChain) maintainTxIndex() {
...
 	defer sub.Unsubscribe()
 
+	// Launch the initial processing if chain is not empty. This step is
+	// useful in these scenarios that chain has no progress and indexer
+	// is never triggered.
+	head := rawdb.ReadHeadBlock(bc.db)
+	tail := rawdb.ReadTxIndexTail(bc.db)
+	if head != nil && tail != nil {
+		done = make(chan struct{})
+		bc.indexBlocks(tail, head.NumberU64(), done)
```
Comment (fjl) [inline core/blockchain.go:2404]: missing `go` here

File: core/blockchain.go [hunk 4]
```diff
@@ -2394,6 +2394,16 @@ func (bc *BlockChain) maintainTxIndex() {
 	}
 	defer sub.Unsubscribe()
 
+	// Launch the initial processing if chain is not empty. This step is
+	// useful in these scenarios that chain has no progress and indexer
+	// is never triggered.
+	head := rawdb.ReadHeadBlock(bc.db)
+	tail := rawdb.ReadTxIndexTail(bc.db)
+	if head != nil && tail != nil {
```
Comment (fjl) [inline core/blockchain.go:2402]: remove `tail` check here

File: core/blockchain.go (hunk 3 above)
Comment (jsvisa) [reply core/blockchain.go:2404]: As discussed with @rjl493456442 we should run it in sync mode.

File: core/blockchain.go (hunk 4 above)
Comment (jsvisa) [reply core/blockchain.go:2402]: If we removed the check here, we need to adjust the txindex test cases in [core/blockchain_test.go](core/blockchain_test.go), because `InsertReceiptChain` will also modify the `TxIndexTail`

https://github.com/ethereum/go-ethereum/blob/4273f248755a48d29002aa6d01de94bca36ea6c1/core/blockchain.go#L1251-L1260


IMO, the purpose of introducing this PR is to continue the indexing/unindexing after the node restarts(even if no new chain is updated)

If `tail` is null, then it means the node is just inited and all blocks are not indexed yet or no blocks were inserted. In this case, We need to wait for at least one new block to be added before we start the indexing/unindexing process.

Let me think about how to modify the test cases if needed.

File: core/blockchain.go (hunk 3 above)
Comment (fjl) [reply core/blockchain.go:2404]: I don't know about this discussion. But if we change it to sync here, it means the process might take a long time to exit when geth is aborted while initial tx indexing is running?

File: core/blockchain.go (hunk 4 above)
Comment (fjl) [reply core/blockchain.go:2402]: I added this comment because @rjl493456442 told me during review, that he thinks it's not necessary to have the `tail` check. You can also leave it in if that's simpler.

File: core/blockchain.go (hunk 3 above)
Comment (jsvisa) [reply core/blockchain.go:2404]: > it means the process might take a long time to exit when geth is aborted while initial tx indexing is running?

I don't think this will happen, as `bc.quit` is directly passed into the `rawdb.IndexTransactions`, `rawdb.UnindexTransactions`, so if `bc.quit` is closed, those processes will fast quit.

File: core/blockchain.go (hunk 4 above)
Comment (jsvisa) [reply core/blockchain.go:2402]: I also didn't think of it before, but the unit test failed and found it possible to pass `tail != nil` check condition, so there is no need to change the unit test.

**Review by rjl493456442 — APPROVED** [review]

File: core/blockchain.go [hunk 5]
```diff
@@ -2394,6 +2400,15 @@ func (bc *BlockChain) maintainTxIndex() {
 	}
 	defer sub.Unsubscribe()
 
+	// Launch the initial processing if chain is not empty. This step is
+	// useful in these scenarios that chain has no progress and indexer
+	// is never triggered.
+	head := rawdb.ReadHeadBlock(bc.db)
+	if head != nil {
```
Comment (holiman) [inline core/blockchain.go:2407]:
```suggestion
	if head := rawdb.ReadHeadBlock(bc.db); head != nil {
```

**Review by holiman — APPROVED** [review]
LGTM
